code/modules/fabrication/fabricator_ui.dm

```diff
@@ -1,94 +1,211 @@
+			continue
+		LAZYADD(build_options, list(ui_fabricator_build_option_entry_data(R)))
+	return build_options
+
+//Fill out the data for a single build option
+/obj/machinery/fabricator/proc/ui_fabricator_build_option_entry_data(var/datum/fabricator_recipe/R)
+	var/list/build_option   = list()
+	var/max_sheets          = (!length(R.resources)) ? 100 : 0
+	var/list/material_costs = ui_fabricator_build_option_cost_list(R, max_sheets)
+
+	build_option["name"]       = R.name
+	build_option["reference"]  = "\ref[R]"
+	build_option["illegal"]    = R.hidden
+
+	if(material_costs)
+		build_option["unavailable"] = !(material_costs["available"])
+		var/list/mats = material_costs["materials"]
+		build_option["materials"]  = length(mats) > 0? mats : null
+
+	build_option["multiplier"] = ui_fabricator_build_option_entry_multiplier_data(R, max_sheets)
+	return build_option
+
+//Returns a list containing a boolean "available" to determine if we can build the recipe, and a list of resources costs ()
+/obj/machinery/fabricator/proc/ui_fabricator_build_option_cost_list(var/datum/fabricator_recipe/R, var/max_sheets)
+	//Make sure it's buildable and list required resources.
+	var/list/material_components = list()
+	var/has_missing_resource = FALSE
+	for(var/material_path in R.resources)
+		var/required_amount = round(R.resources[material_path] * mat_efficiency)
+		var/sheets          = round(stored_material[material_path] / required_amount)
+		var/has_enough      = TRUE
+
+		if(max_sheets == 0 || max_sheets < sheets)
+			max_sheets = sheets
+		if(stored_material[material_path] < required_amount)
+			has_missing_resource = TRUE
+			has_enough = FALSE
+
+		//Must make it a double list here or the fields are just overwriting eachothers
+		material_components += list(list(
+				"name"       = stored_substances_to_names[material_path],
+				"amount"     = required_amount,
+				"has_enough" = has_enough, 
+			))
+	return list("available" = !has_missing_resource && ui_fabricator_build_option_is_available(R, max_sheets), "materials" = material_components)
 
+//Override to add more checks to make a build option unavailable. EX: if the machine requires a setting to be set first
+/obj/machinery/fabricator/proc/ui_fabricator_build_option_is_available(var/datum/fabricator_recipe/R, var/max_sheets)
+	return TRUE
+
+/obj/machinery/fabricator/proc/ui_fabricator_build_option_entry_multiplier_data(var/datum/fabricator_recipe/R, var/max_sheets)
+	var/list/multiplier
+	if(R.max_amount >= PRINT_MULTIPLIER_DIVISOR && max_sheets >= PRINT_MULTIPLIER_DIVISOR)
+		multiplier = list()
+		for(var/i = 1 to FLOOR(min(R.max_amount, max_sheets)/PRINT_MULTIPLIER_DIVISOR))
+			var/mult = i * PRINT_MULTIPLIER_DIVISOR
+			multiplier += list(list("label" = "x[mult]", "multiplier" = mult))
+	return multiplier
+
+//
+// UI sections data
+//
+/obj/machinery/fabricator/ui_data(mob/user, ui_key)
+	var/list/data = ..()
+	//Common fab data
+	data += ui_data_status(user, ui_key)
```

Comment:
There are actually small issues with doing this: these procs return lazylists (or some of them do, at any rate) so they might return `null`, which here then adds a `null` literal to the list; this will then lead to potential parsing errors in the front end. Make sure they are returning an actual, possibly empty, list, not a lazy list.

Reply:
Good catch!

Reply:
fixed